### DISCUSSION TITLE
Getting number of updated rows with 2.0-style queries?

### DISCUSSION THREAD

**Discussion by ljiatu** [sqlalchemy/sqlalchemy#6167] [Usage Questions] [answered]
The "bulk update" query in 1.x-style returns the number of updated rows. However, the 2.0-style query doesn't do that anymore. Here's a demonstration:

```
@pytest.mark.asyncio
async def test(db_session: Session, async_db_session: AsyncSession, db_items: List[Item]) -> None:
    db_session.add_all(db_items)
    db_session.commit()

    num_updated = db_session.query(Item).update({"show_in_gallery": True}, synchronize_session=False)
    print(num_updated)  # returns 3
    async_result = await async_db_session.execute(update(Item).values(show_in_gallery=True).execution_options(synchronize_session=False))
    async_num_updated = async_result.scalar()  # raises exception
```

`db_session` is a regular synchronous session, `async_db_session` is an async session. `db_items` is a list of `Item` objects that have a `show_in_gallery` field. The last line of code raises an exception `sqlalchemy.exc.ResourceClosedError: This result object does not return rows. It has been closed automatically.` (understandably). So what is the 2.0-style equivalent way of retrieving the number of affected rows?

Thanks!

**Comment by zzzeek** [comment, marked as answer]
session.execute() returns a [CursorResult](https://docs.sqlalchemy.org/en/14/core/connections.html?highlight=cursorresult#sqlalchemy.engine.CursorResult) object in the case of DML statements such as UPDATE and DELETE.   This is the result that came right from the connection.execute().   you use the [.rowcount](https://docs.sqlalchemy.org/en/14/core/connections.html?highlight=cursorresult%20rowcount#sqlalchemy.engine.CursorResult.rowcount) attribute to get the number of rows affected.

**Reply by zzzeek** [reply]
ill add this to the docs for the ORM update/delete section.

**Reply by ljiatu** [reply]
Cool thank you!

**Reply by ljiatu** [reply]
Hmm do I directly access rowcount like `async_num_updated.rowcount`? I just tried it and it returned 0, but `num_updated` is 3.

**Reply by ljiatu** [reply]
I'm using the `ascynpg` driver, and `supports_sane_rowcount` returns true.

**Reply by zzzeek** [reply]
you'd need to provide an MCVE.  Try the example script below and show me how to make it fail:

```python
import asyncio

from sqlalchemy import Column
from sqlalchemy import Integer
from sqlalchemy import String
from sqlalchemy import update
from sqlalchemy.ext.asyncio import AsyncSession
from sqlalchemy.ext.asyncio import create_async_engine
from sqlalchemy.ext.declarative import declarative_base
from sqlalchemy.orm import sessionmaker

Base = declarative_base()


class A(Base):
    __tablename__ = "a"

    id = Column(Integer, primary_key=True)
    data = Column(String)


async def async_main():
    engine = create_async_engine(
        "postgresql+asyncpg://scott:tiger@localhost/test",
        echo=True,
    )

    async with engine.begin() as conn:
        await conn.run_sync(Base.metadata.drop_all)
        await conn.run_sync(Base.metadata.create_all)

    async_session = sessionmaker(
        engine, class_=AsyncSession
    )

    async with async_session() as session:
        async with session.begin():
            session.add_all([A(data="a1"), A(data="a2"), A(data="a3")])

        result = await session.execute(
            update(A)
            .where(A.data.in_(["a1", "a2", "a3"]))
            .values(data="a4")
            .execution_options(synchronize_session=False)
        )
        assert result.rowcount == 3


asyncio.run(async_main())

```

**Reply by ljiatu** [reply]
🤦  sorry it was my mistake. `rowcount` works. I was not using the correct session. Thanks again for your help!
